File: AndroidApp/src/main/java/com/otero/recipetoshop/android/presentation/controllers/listacompra/ListaCompraViewModel.kt
package com.otero.recipetoshop.android.presentation.controllers.listacompra

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import com.otero.recipetoshop.presentationlogic.states.listacompra.ListaCompraState
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import androidx.lifecycle.viewModelScope
import com.otero.recipetoshop.Interactors.listacompra.*
import com.otero.recipetoshop.domain.model.ListaCompra.Productos
import com.otero.recipetoshop.domain.model.ListaCompra.toAlimento
import com.otero.recipetoshop.domain.model.despensa.Alimento
import com.otero.recipetoshop.domain.model.despensa.toProductos
import com.otero.recipetoshop.domain.dataEstructres.FilterEnum
import com.otero.recipetoshop.domain.dataEstructres.SupermercadosEnum
import com.otero.recipetoshop.presentationlogic.events.listacompra.ListaCompraEvents
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch


@HiltViewModel
class ListaCompraViewModel
@Inject
constructor(
    savedStateHandle: SavedStateHandle,
    private val calcularProductos: CalcularProductos,
    private val getProductos: ObtenerProductos,
    private val deleteProductos: DeleteProductos,
    private val getAlimentosNoEncontradosCache: GetAlimentosNoEncontradosCache,
    private val saveListaCompra: SaveListaCompra,
    private val updateProducto: UpdateProducto,
    private val finalizarCompra: FinalizarCompra
):ViewModel(){
    //Estado de la lista de la compra
    val listaCompraState: MutableState<ListaCompraState> = mutableStateOf(ListaCompraState())

    init {
        try {
            //Obtengo el id de la cesta de la compra a tarnsformar
            savedStateHandle.get<Int>("cestacompraid")?.let { cestacompraid ->
                viewModelScope.launch {
                    listaCompraState.value = listaCompraState.value.copy(id_cestaCompra = cestacompraid)
                }
            }
        } catch(e: Exception){
            println("No se ha podido obtener el identificador de la lista de recetas: " + e.message)
        }

        //Si hemos llegado directos del menú de navegación, entonces compruebo si hay lista calculada, sino folio en blanco de momento.
        if(listaCompraState.value.id_cestaCompra == -1){
            //Lo sé porque al navegar hasta aquí guardo en savedStateHandle -1 como id de lista de la compra.
            actualizaLista()
        }
        //En caso contrario estamos llegando desde el cálculo de una cesta de la compra, hay que calcularlos.
        else {
            //Elimino de cache la lista anterior si la hubiese.
            deleteProductosCache()
            //Actualizo la lista de la compra con los productos calculados.
            calcularProductos()
        }
    }

    fun onTriggerEvent(event: ListaCompraEvents){
        when(event){
            is ListaCompraEvents.onCLickFilter -> {
                //Actualizo la lsita de la compra según el filter aplicado
                aplicarFlitro(FilterEnum.parseString(event.filter_nombre))
            }

            is ListaCompraEvents.onClickProducto -> {
                actualizarProducto(active = !event.producto.active, producto = event.producto)
            }

            is ListaCompraEvents.onFinishCompra -> {
                finalizarCompra()
            }
            else -> {
                print("Evento de la lista de la compra no esperado.")
            }
        }
    }

    private fun finalizarCompra() {
        finalizarCompra.finalizarCompra(
            id_cestaCompra = listaCompraState.value.id_cestaCompra,
            productos = ArrayList(listaCompraState.value.listaProductos)
        ).onEach { dataState ->
        }.launchIn(viewModelScope)
    }

    private fun actualizarProducto(active: Boolean, producto: Productos.Producto) {
        updateProducto.updateProducto(active = active, producto = producto).onEach { dataState ->
            //Actualizo la lista de la compra
            actualizaLista()
        }.launchIn(viewModelScope)
    }

    private fun aplicarFlitro(filter: FilterEnum) {
        //Seteo la lista de la compra a los productos anteriores.
        var id_anterior = -1
        if(listaCompraState.value.listaProductos.isNotEmpty()){
            id_anterior = listaCompraState.value.listaProductos.first().id_cestaCompra
        } else if(listaCompraState.value.alimentos_no_encontrados.isNotEmpty()) {
            id_anterior = listaCompraState.value.alimentos_no_encontrados.first().id_cestaCompra!!
        }
        listaCompraState.value = listaCompraState.value.copy(id_cestaCompra = id_anterior)
        //Elimino los productos calculados hasta le momento
        deleteProductosCache()
        //Calculo los productos con el filtro seleccionado
        calcularProductos(filter = filter)
    }

    private fun deleteProductosCache() {
        deleteProductos.deleteProductos().onEach { DataState ->
            DataState.data?.let {
                //Productos eliminados
            }
        }.launchIn(viewModelScope)
    }

    private fun actualizaLista() {
        //TODO: Falta que los supermercados se puedan elegir dinámicamente
        listaCompraState.value = listaCompraState.value.copy(
            supermercados = mutableSetOf(
                SupermercadosEnum.CARREFOUR,
                SupermercadosEnum.MERCADONA,
                SupermercadosEnum.DIA
            )
        )
        //Obtengo y actualizo productos encontrados.
        getProductosEncontrados()
        //Obtengo y actualizo productos no encontrados.
        getProductosNoEncontrados()
    }

    private fun getProductosNoEncontrados(){
        getAlimentosNoEncontradosCache.getAlimentosNoEncontradosCache().onEach {DataState ->
            DataState.data?.let { productosNoEncontrados ->
               if(!productosNoEncontrados.isEmpty()){
                   //Actualizo la lista de la compra con los productos encontrados en cache.
                   listaCompraState.value = listaCompraState.value.copy(
                       //Devuelvo la lista de los productos no encontrados como alimentos con ID -1.
                       alimentos_no_encontrados = productosNoEncontrados.map{ producto ->
                           producto.toAlimento().copy(id_alimento = -1)
                       },
                   )
               }
            }
        }.launchIn(viewModelScope)
    }

    private fun getProductosEncontrados() {
        getProductos.obtenerProductos().onEach {DataState ->
            DataState.data?.let { productosEncontrados ->
                if(!productosEncontrados.productos_cache.isEmpty()){
                    //Actualizo la lista de la compra con los productos no encontrados en cache.
                    listaCompraState.value = listaCompraState.value.copy(
                        listaProductos = productosEncontrados.productos_cache,
                        id_cestaCompra = productosEncontrados.id_cestaCompra
                    )
                    //Actualizo el precio y peso totales de la lista de la compra-
                    actualizarListaProductosEncontrados(listaCompraState.value.listaProductos)
                }
            }
        }.launchIn(viewModelScope)
    }

    private fun calcularProductos(filter: FilterEnum = FilterEnum.BARATOS, supermercadosEnum: SupermercadosEnum = SupermercadosEnum.CARREFOUR) {
        //TODO: Falta que los supermercados se puedan elegir dinámicamente
        listaCompraState.value = listaCompraState.value.copy(
            supermercados = mutableSetOf(
                SupermercadosEnum.CARREFOUR,
                SupermercadosEnum.MERCADONA,
                SupermercadosEnum.DIA
            )
        )

        //Calculo los productos.
        calcularProductos.calcularProductos(
            id_cestaCompra = listaCompraState.value.id_cestaCompra,
            supermercados = listaCompraState.value.supermercados,
            filter = filter
        ).onEach { dataState ->
            dataState.data?.let { alimentos_productosEncontrados ->
                //Actualizo el precio total y el peso total
                actualizarListaProductosEncontrados(productosEncontrados = alimentos_productosEncontrados.second)
                //Calculo los alimentos no encontrados.
                calcularAlimentosNoEncontrados(
                    alimentosNecesarios = alimentos_productosEncontrados.first,
                    productos_encontrados = alimentos_productosEncontrados.second
                )
                //Guardo los productos encontrados en cache.
                saveProductosEncontrados(productos_encontrados = alimentos_productosEncontrados.second)
                //Guardo los productos no encontrados en cache.
                saveProductosNoEncontrados()

            }
        }.launchIn(viewModelScope)
    }

    private fun saveProductosNoEncontrados() {
        //Tranformo los productos no encontrados en alimentos.
        val productosNoEncontrados = listaCompraState.value.alimentos_no_encontrados.toProductos()
        //Guardo en cache los alimentos no enocntrados
        saveListaCompra.saveAlimentosNoEncontrados(
            id_cestaCompra = listaCompraState.value.id_cestaCompra,
            productosNoEncontrados = productosNoEncontrados
        ).onEach { dataState ->
            dataState.data?.let {
                //Productos no encontrados guardados
                //Actualizo la lista.
                actualizaLista()
            }
        }.launchIn(viewModelScope)
    }

    private fun saveProductosEncontrados(productos_encontrados: List<Productos.Producto>) {
        //Guardo en cache los productos encontrados de la cesta de la compra actual.
        saveListaCompra.saveProductos(
            id_cestaCompra = listaCompraState.value.id_cestaCompra,
            productos = productos_encontrados
        ).onEach { dataState ->
            dataState.data?.let {
                //Productos encontrados guardados
            }
        }.launchIn(viewModelScope)
    }

    private fun actualizarListaProductosEncontrados(productosEncontrados: List<Productos.Producto>) {

        //Actualizo los supermercados seleccionados y con productos encontrados.
        actualziarSupermercados(productosEncontrados)

        //Actualizo el precio total y atualizo el peso total de los productos (considero que los ml pesan igual que los Kg).
        var precio_total_carrefour = 0f
        var precio_total_dia = 0f
        var precio_total_mercadona = 0f

        var peso_total_carrefour = 0f
        var peso_total_dia = 0f
        var peso_total_mercadona = 0f

        for(producto in productosEncontrados){
            when(producto.supermercado){
                SupermercadosEnum.CARREFOUR -> {
                    precio_total_carrefour += producto.cantidad * producto.precio_numero
                    if((producto.tipoUnidad != null && !producto.tipoUnidad!!.name.equals("UNIDADES"))){
                        peso_total_carrefour += producto.cantidad * producto.peso
                    }
                }
                SupermercadosEnum.MERCADONA -> {
                    precio_total_mercadona += producto.cantidad * producto.precio_numero
                    if((producto.tipoUnidad != null && !producto.tipoUnidad!!.name.equals("UNIDADES"))){
                        peso_total_mercadona += producto.cantidad * producto.peso
                    }
                }
                SupermercadosEnum.DIA -> {
                    precio_total_dia += producto.cantidad * producto.precio_numero
                    if((producto.tipoUnidad != null && !producto.tipoUnidad!!.name.equals("UNIDADES"))){
                        peso_total_dia += producto.cantidad * producto.peso
                    }
                }
            }
        }

        //Los actualizo.
        listaCompraState.value = listaCompraState.value.copy(
            precio_total = hashMapOf(
                SupermercadosEnum.CARREFOUR to precio_total_carrefour,
                SupermercadosEnum.DIA to precio_total_dia,
                SupermercadosEnum.MERCADONA to precio_total_mercadona
            ),
            peso_total = hashMapOf(
                SupermercadosEnum.CARREFOUR to peso_total_carrefour,
                SupermercadosEnum.DIA to peso_total_dia,
                SupermercadosEnum.MERCADONA to peso_total_mercadona
            )
        )
    }

    //Metodo para calcular los alimentos que no han sido encontrados o que ha habido problemas en calcular las cantidades.
    private fun calcularAlimentosNoEncontrados(alimentosNecesarios: ArrayList<Alimento>, productos_encontrados: List<Productos.Producto>) {
        val alimentosNoEncontrados: ArrayList<Alimento> = arrayListOf()
        for(alimento in alimentosNecesarios){
            var match: Boolean = false
            for(producto in productos_encontrados){
                //Si encuentra al menos un producto de su tipo; sí ha sido encontrado.
                if(producto.query.trim().lowercase().equals(alimento.nombre.trim().lowercase())){
                    match = true
                }
            }
            //Si no se ha encontrado ningún producto con dicho alimento, este se añade a los alimentos np encontrados a la lista de la compra.
            if(!match){
                alimentosNoEncontrados.add(alimento)
            }
        }
        //Actualizo la lista de alimentos no encontrados en el state de la lista de la compra.
        listaCompraState.value = listaCompraState.value.copy(alimentos_no_encontrados = alimentosNoEncontrados)
    }

    private fun actualziarSupermercados(productos: List<Productos.Producto>){
        val supermercados_actuales: MutableSet<SupermercadosEnum> = mutableSetOf()
        for(supermercado in SupermercadosEnum.values()){
            if (
                productos.any{ producto ->
                    producto.supermercado.equals(supermercado)
                }
            ){
                supermercados_actuales.add(supermercado)
            }
        }
        listaCompraState.value = listaCompraState.value.copy(supermercados = supermercados_actuales)
    }
}

File: IOSApp/IOSApp/Presentation/ViewModels/ListaCompraViewModel.swift
//
//  ListaCompraViewModel.swift
//  IOSApp
//
//  Created by David Otero Diaz on 8/7/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import sharedApp

class ListaCompraViewModel: ObservableObject {
    
    private let caseUses: UseCases
    private let id_listaRecetas: Int
    
    @Published var state: ListaCompraState = ListaCompraState()
    
    init(
        caseUses: UseCases,
        id_listaRecetas: Int
    ){
        print("ID lista de la compra: " + String(id_listaRecetas))
        
        self.caseUses = caseUses
        self.id_listaRecetas = id_listaRecetas
        
        //Primero hay que eliminar la lista anterior.
        self.deleteListaCompra()
        
        if(self.id_listaRecetas != -1){
            //TODO: Falta que se elijan los supermercados dinámicamente.
            self.state = self.state.doCopy(
                listaProductos: self.state.listaProductos,
                id_cestaCompra: Int32(self.id_listaRecetas),
                precio_total: self.state.precio_total,
                peso_total: self.state.peso_total,
                alimentos_cesta: self.state.alimentos_cesta,
                alimentos_no_encontrados: self.state.alimentos_no_encontrados,
                supermercados: KotlinMutableSet<SupermercadosEnum>(array: [
                    SupermercadosEnum.carrefour,
                    SupermercadosEnum.dia,
                    SupermercadosEnum.mercadona
                ])
            )
            print("Los supermercados son: ")
            print(self.state.supermercados.count)
            dump(self.state.supermercados)
            self.aplicarFiltro(filter: FilterEnum.baratos)
        }
    }
    
    func onTriggerEvent(event: ListaCompraEvents) -> Void {
        switch event {
            case is ListaCompraEvents.onCLickFilter:
                self.aplicarFiltro(
                    filter: ParserFilterEnum().parseStringToFilter(
                        nombre_filtro: (event as! ListaCompraEvents.onCLickFilter).filter_nombre
                    )
                )
            case is ListaCompraEvents.onClickProducto:
                print("He llegado a intentar actualizar")
                print((event as! ListaCompraEvents.onClickProducto).producto.active)
                self.updateProducto(
                    active: !(event as! ListaCompraEvents.onClickProducto).producto.active,
                    producto: (event as! ListaCompraEvents.onClickProducto).producto
                )
            case is ListaCompraEvents.onFinishCompra:
                self.finalizarCompra()
            default:
                print("La lista de la compra no está preparada para el evento")
        }
    }
    
    private func finalizarCompra() -> Void{
        self.caseUses.finalizarCompra.finalizarCompra(
            id_cestaCompra: self.state.id_cestaCompra,
            productos: NSMutableArray(array: self.state.listaProductos)
        ).collectFlow(
            coroutineScope: nil,
            callback: { dataState in
                //Navegar a la pantalla de despensa.
                if(dataState?.data != nil){
                    self.deleteListaCompra()
                }
            })
    }
    
    private func actualizarLista() -> Void{
        //Primero obtengo los productos en caché
        self.caseUses.getProductos.obtenerProductos().collectFlow(
            coroutineScope: nil,
            callback: { dataState in
                if(dataState?.data != nil){
                    //Si se han encontrado productos
                    if(!(dataState?.data?.productos_cache.isEmpty)!){
                        print("Los productos obtenidos son: ")
                        dump(dataState?.data?.productos_cache)
                        self.state = self.state.doCopy(
                            listaProductos: (dataState?.data!.productos_cache)!,
                            id_cestaCompra: (dataState?.data!.id_cestaCompra)!,
                            precio_total: self.state.precio_total,
                            peso_total: self.state.peso_total,
                            alimentos_cesta: self.state.alimentos_cesta,
                            alimentos_no_encontrados: self.state.alimentos_no_encontrados,
                            supermercados: self.state.supermercados
                        )
                    }
                    
                    //Obtengo los alimentos no encontrados en caché
                    self.caseUses.getAlimentosNoEncontradosCache.getAlimentosNoEncontradosCache().collectFlow(
                        coroutineScope: nil,
                        callback: { productosNoEncontrados in
                            if(productosNoEncontrados?.data != nil){
                                self.state = self.state.doCopy(
                                    listaProductos: self.state.listaProductos,
                                    id_cestaCompra: self.state.id_cestaCompra,
                                    precio_total: self.state.precio_total,
                                    peso_total: self.state.peso_total,
                                    alimentos_cesta: self.state.alimentos_cesta,
                                    alimentos_no_encontrados: (
                                        productosNoEncontrados?.data?.map{ producto in
                                            (producto as! Productos.Producto).toAlimento().doCopy(
                                                id_cestaCompra:  ((producto as! Productos.Producto).id_cestaCompra as! KotlinInt),
                                                id_receta:  -1,
                                                id_alimento: -1,
                                                nombre:  (producto as! Productos.Producto).nombre,
                                                cantidad:  Int32((producto as! Productos.Producto).peso),
                                                tipoUnidad:  (producto as! Productos.Producto).tipoUnidad!,
                                                active: true
                                            )
                                        }
                                    )as! [Alimento],
                                    supermercados: self.state.supermercados)
                            }
                            //Actualizo los pesos y precios totales.
                            self.actualizarPesosYPreciosTotales()
                        }
                    )
                }
            }
        )
    }
    
    private func updateProducto(active: Bool, producto: Productos.Producto) -> Void {
        caseUses.updateProducto.updateProducto(active: active, producto: producto).collectFlow(
            coroutineScope: nil,
            callback: { dataState in
                if(dataState?.data != nil){
                    //Actualizo la lista
                    self.actualizarLista()
                }
            })
    }
    
    private func aplicarFiltro(filter: FilterEnum) -> Void{
        //Elimino los productos calculados hasta ahora
        deleteListaCompra()
        
        print("Aplicando filtrado de tipo: ")
        print(filter.name)
        //Calculo los productos con el nuevo filtro
        calcularCompra(
            filter: filter,
            supermercados: KotlinMutableSet<SupermercadosEnum>(array: [
                SupermercadosEnum.carrefour,
                SupermercadosEnum.dia,
                SupermercadosEnum.mercadona
            ])
        )
        
    }
    
    private func calcularCompra(
        filter: FilterEnum = FilterEnum.baratos,
        supermercados: KotlinMutableSet<SupermercadosEnum> = KotlinMutableSet<SupermercadosEnum>(array: [SupermercadosEnum.carrefour])
    ) -> Void {
        var productos: [Productos.Producto] = []
        var alimentosNecesarios: [Alimento] = []
        
        //Primero calculo los mejores productos. Obtengo a la vez los alimentos necesarios
        caseUses.calcularProductos.calcularProductos(
            id_cestaCompra: self.state.id_cestaCompra,
            filter: filter,
            supermercados: supermercados 
        ).collectFlow(
            coroutineScope: nil,
            callback: { dataState in
                if dataState?.data != nil {
                    productos = dataState?.data?.second as! [Productos.Producto]
                    alimentosNecesarios = dataState?.data?.first as! [Alimento]
                    print("Productos de la compra de la lista con id " + String(self.id_listaRecetas) + ": "
                          + String(productos.count))
                    dump(productos)
                    //Guardo los productos encontrados en caché.
                    self.guardarProductosCache(productos: productos)
                    
                    //Guardo los productos no encontrados en caché.
                    self.calcularAlimentosNoEncontrados(alimentosNecesarios: alimentosNecesarios,productos: productos)
                    self.guardarProductosNoEncontradosCache()
                    
                    //Lo anterior guarda los productos no encontrados y actualiza la lista y los pesos y precios totales.
                }
            })
    }
    
    private func actualizarSupermercados(productos: [Productos.Producto]) -> Void{
        let supermercados_actuales = KotlinMutableSet<SupermercadosEnum>()
        let iterator = SupermercadosEnum.values().iterator()
        while(iterator.hasNext()){
            let supermercado = iterator.next_()
            if(productos.contains(where: { producto in
                producto.supermercado.isEqual(supermercado)
            })){
                //Añado el supermercado a los supermercados actuales.
                supermercados_actuales.add(supermercado as! SupermercadosEnum)
            }
        }
        self.state = self.state.doCopy(
            listaProductos: self.state.listaProductos,
            id_cestaCompra: self.state.id_cestaCompra,
            precio_total: self.state.precio_total,
            peso_total: self.state.peso_total,
            alimentos_cesta: self.state.alimentos_cesta,
            alimentos_no_encontrados: self.state.alimentos_no_encontrados,
            supermercados: supermercados_actuales
        )
    }
    
    private func actualizarPesosYPreciosTotales() -> Void{
        
        //Actualizo los supermercados
        self.actualizarSupermercados(productos: self.state.listaProductos)
        
        //Creo las variable necesarias
        var precio_total_carrefour: Float = 0
        var precio_total_dia: Float = 0
        var precio_total_mercadona: Float = 0

        var peso_total_carrefour: Float = 0
        var peso_total_dia: Float = 0
        var peso_total_mercadona: Float = 0
        
        print("Antes de calcular los pesos totales")
        dump(self.state.listaProductos)
        //Calculo los precios totales y pesos totales de cada supermercado
        self.state.listaProductos.forEach{ producto in
            switch(producto.supermercado){
                case SupermercadosEnum.carrefour:
                    precio_total_carrefour += Float(producto.cantidad) * producto.precio_numero
                    if((producto.tipoUnidad != nil && !(producto.tipoUnidad!.name == "UNIDADES"))){
                        peso_total_carrefour += Float(producto.cantidad) * producto.peso
                    }
                case SupermercadosEnum.mercadona:
                    precio_total_mercadona += Float(producto.cantidad) * producto.precio_numero
                    if((producto.tipoUnidad != nil && !(producto.tipoUnidad!.name == "UNIDADES"))){
                        peso_total_mercadona += Float(producto.cantidad) * producto.peso
                    }
                case SupermercadosEnum.dia:
                    precio_total_dia += Float(producto.cantidad) * producto.precio_numero
                    if((producto.tipoUnidad != nil && !(producto.tipoUnidad!.name == "UNIDADES"))){
                        peso_total_dia += Float(producto.cantidad) * producto.peso
                    }
                default:
                    print("SUpermercado no filtrado correctamente en cantidades y precios")
            }
        }
        
        let dictionary_precio = KotlinMutableDictionary<SupermercadosEnum, KotlinFloat>()
        dictionary_precio[SupermercadosEnum.carrefour] = precio_total_carrefour
        dictionary_precio[SupermercadosEnum.mercadona] = precio_total_mercadona
        dictionary_precio[SupermercadosEnum.dia] = precio_total_dia
        
        let dictionary_peso = KotlinMutableDictionary<SupermercadosEnum, KotlinFloat>()
        dictionary_peso[SupermercadosEnum.carrefour] = peso_total_carrefour
        dictionary_peso[SupermercadosEnum.mercadona] = peso_total_mercadona
        dictionary_peso[SupermercadosEnum.dia] = peso_total_dia
        
        //Actualizo el estado con los productos y el nuevo precio y peso total.
        let currentState: ListaCompraState = self.state.copy() as! ListaCompraState
        self.state = self.state.doCopy(
            listaProductos: currentState.listaProductos,
            id_cestaCompra: currentState.id_cestaCompra,
            precio_total: dictionary_precio ,
            peso_total: dictionary_peso ,
            alimentos_cesta: currentState.alimentos_cesta,
            alimentos_no_encontrados: currentState.alimentos_no_encontrados,
            supermercados: currentState.supermercados
        )
    }
    
    private func calcularAlimentosNoEncontrados(alimentosNecesarios: [Alimento], productos: [Productos.Producto]) -> Void {
        var alimentosNoEncontrados: [Alimento] = []
        alimentosNecesarios.forEach{ alimento in
            if(!productos.contains(
                where: { producto in
                    producto.query.lowercased() == alimento.nombre.lowercased()
                }
            )){
                //Añado el alimento a la lista de alimentos no encontrados.
                alimentosNoEncontrados.append(alimento)
            }
        }
        //Guardo los alimentos no encontrados en el estado.
        let currentState: ListaCompraState = self.state.copy() as! ListaCompraState
        self.state = self.state.doCopy(
            listaProductos: currentState.listaProductos,
            id_cestaCompra: currentState.id_cestaCompra,
            precio_total: currentState.precio_total,
            peso_total: currentState.peso_total,
            alimentos_cesta: currentState.alimentos_cesta,
            alimentos_no_encontrados: alimentosNoEncontrados,
            supermercados: currentState.supermercados
        )
    }
    
    private func guardarProductosCache(productos: [Productos.Producto]) -> Void {
        caseUses.saveListaCompra.saveProductos(
            id_cestaCompra: self.state.id_cestaCompra,
            productos: productos
        ).collectFlow(
            coroutineScope: nil,
            callback: { dataState in
                //Productos guardados
            })
    }
    
    private func guardarProductosNoEncontradosCache() -> Void {
        caseUses.saveListaCompra.saveAlimentosNoEncontrados(
            id_cestaCompra: self.state.id_cestaCompra,
            productosNoEncontrados: state.alimentos_no_encontrados.toProductos()
        ).collectFlow(
            coroutineScope: nil,
            callback: { dataState in
                //Alimentos no encontrados guardados.
                self.actualizarLista()
            }
        )
    }
    
    private func deleteListaCompra() -> Void {
        caseUses.deleteProductos.deleteProductos().collectFlow(
            coroutineScope: nil,
            callback: { dataState in
                if dataState?.data != nil {
                    //Productos guardados eliminados.
                    self.state = self.state.doCopy(
                        listaProductos: [],
                        id_cestaCompra: Int32(self.id_listaRecetas),
                        precio_total: KotlinMutableDictionary<SupermercadosEnum, KotlinFloat>(),
                        peso_total: KotlinMutableDictionary<SupermercadosEnum, KotlinFloat>() ,
                        alimentos_cesta: [],
                        alimentos_no_encontrados: [],
                        supermercados: KotlinMutableSet<SupermercadosEnum>()
                    )
                }
            }
        )
    }
}
